consensus/istanbul/core/handler.go

```diff
@@ -174,43 +248,226 @@ func (c *core) handleMsg(payload []byte) error {
 	}
 
 	// Only accept message if the address is valid
-	_, src := c.current.ValidatorSet().GetByAddress(msg.Address)
-	if src == nil {
+	_, validator := c.current.ValidatorSet().GetByAddress(msg.Address)
+	if validator == nil {
 		logger.Error("Invalid address in message", "m", msg)
 		return istanbul.ErrUnauthorizedAddress
 	}
 
-	return c.handleCheckedMsg(msg, src)
-}
+	// Update logger context
+	logger = logger.New("from", msg.Address, "curr_view", c.current.View())
 
-func (c *core) handleCheckedMsg(msg *istanbul.Message, src istanbul.Validator) error {
-	logger := c.newLogger("func", "handleCheckedMsg", "from", msg.Address)
-
-	// Store the message if it's a future message
-	catchFutureMessages := func(err error) error {
-		if err == errFutureMessage {
-			// Store in backlog (if it's not from self)
-			if msg.Address != c.address {
-				c.backlog.store(msg)
+	m, err := buildAlgorithmMsg(msg, c.oracle)
```

Comment:
It is effectively stateless, this is because although the oracle is long lived the results of calling `SetRoundChangeCertificate` are only ever referenced when processing the preprepare message that resulted in the call to `SetRoundChangeCertificate`. This is easily verifiable because we can see that `ValidRoundChangeCert` is only called for preprepare messages (this is the only reference in the code).
https://github.com/celo-org/celo-blockchain/blob/baa55bd4837d3690c03324429d1858164b45a7de/consensus/istanbul/algorithm/algorithm.go#L147

And any subsequent preprepare message will overwrite the prior results.
https://github.com/celo-org/celo-blockchain/blob/baa55bd4837d3690c03324429d1858164b45a7de/consensus/istanbul/core/handler.go#L202-L204

It's not my ideal approach, but given the time constraints this seemed like the most pragmatic option. I don't see it being too much work to properly represent the round change cert in an `algorithm.Msg` which would then make this process obviously stateless.